arangod/Cluster/RebootTracker.cpp

```diff
@@ -66,296 +85,148 @@ RebootTracker::RebootTracker(RebootTracker::SchedulerPointer scheduler)
+    queueCallback({std::move(callback), std::string{description}});
+    return CallbackGuard{};
   }
 
   // For the given server, get the existing rebootId => [callbacks] map,
   // or create a new one
-  auto& rebootIdMap = _callbacks[peerState.serverId()];
-  // For the given rebootId, get the existing callbacks map,
-  // or create a new one
-  auto& callbackMapPtr = rebootIdMap[peerState.rebootId()];
-
-  if (callbackMapPtr == nullptr) {
-    // We must never leave a nullptr in here!
-    // Try to create a new map, or remove the entry.
-    try {
-      callbackMapPtr = std::make_shared<std::remove_reference<
-          decltype(callbackMapPtr)>::type::element_type>();
-    } catch (...) {
-      rebootIdMap.erase(peerState.rebootId());
-      throw;
-    }
-  }
-
-  TRI_ASSERT(callbackMapPtr != nullptr);
-
-  auto& callbackMap = *callbackMapPtr;
-
-  auto const callbackId = getNextCallbackId();
-
-  // The guard constructor might, theoretically, throw, and so can constructing
-  // the std::function. So we need to construct it before emplacing the
-  // callback.
-  auto callbackGuard = CallbackGuard([this, peerState, callbackId]() {
-    unregisterCallback(peerState, callbackId);
-  });
-
-  auto const [iterator, inserted] = callbackMap.try_emplace(
-      callbackId,
-      DescriptedCallback{std::move(callback), std::move(callbackDescription)});
+  auto& rebootsMap = _callbacks[serverId];
+  // For the given rebootId, get the existing callbacks map, or create a new one
+  auto& callbacksMap = rebootsMap[rebootId];
+  auto const callbackId = ++_callbackId;
+  CallbackGuard callbackGuard{[this, serverId = std::string{serverId}, rebootId,
+                               callbackId]() noexcept {
+    unregisterCallback(serverId, rebootId, callbackId);
+  }};
+  auto [_, inserted] = callbacksMap.try_emplace(callbackId, std::move(callback),
+                                                std::string{description});
   TRI_ASSERT(inserted);
-  TRI_ASSERT(callbackId == iterator->first);
-
   return callbackGuard;
 }
 
-void RebootTracker::scheduleAllCallbacksFor(ServerID const& serverId) {
-  scheduleCallbacksFor(serverId, RebootId::max());
-  // Now the rebootId map of this server, if it exists, must be empty.
-  TRI_ASSERT(_callbacks.find(serverId) == _callbacks.end() ||
-             _callbacks.find(serverId)->second.empty());
-}
-
-// This function may throw.
-// If (and only if) it returns, it has scheduled all affected callbacks, and
-// removed them from the registry.
-// Otherwise the state is unchanged.
-void RebootTracker::scheduleCallbacksFor(ServerID const& serverId,
-                                         RebootId rebootId) {
-  _mutex.assertLockedByCurrentThread();
-
-  auto serverIt = _callbacks.find(serverId);
-  if (serverIt != _callbacks.end()) {
-    auto& rebootMap = serverIt->second;
-    auto const begin = rebootMap.begin();
-    // lower_bounds returns the first iterator that is *not less than* rebootId
```

Comment:
I'm not confusing:

Even code like 
```
for (auto it = begin; it != end && it->first < rebootId;) {
  ...
}
```
twice

is better than lower_bound + distance + for

Also, I don't understand why you are protecting old code. 
If it bothers you that I changed your code, we can discuss it but not in PR?


>  and in no test case you will ever be able to measure it.

I don't think you right all this code called under mutex, so any other access to reboot tracker will be waiting while we end this operation, this can affect thread pool and other tasks.
Simply talk, slowing down this code is not local.

Reply:
The map contains a single element. One. If another server reboots (or fails), there's a short time window where it _might_ get a second entry. Then this code will be executed _once_, per rebooted (or failed) server. Can we please stop talking about the complexity of it, and about whether it might take longer to iterate twice over that one element?

It should be clear what the code does, and it should be obviously correct. So for example, I'd be perfectly happy with deleting the reserve/std::distance line: It's useless, and just adds noise. If you think you have other ideas that make it more clear, I'm happy to hear them. But performance is not an argument.